### PR TITLE
ENH: static typing for qhull

### LINKED ISSUE #13299: Typing for scipy.spatial
**Is your feature request related to a problem? Please describe.**
It would be great to have typing for spatial (e.g. Voronoi) as some fields are undocumented (e.g. ridge_dict), and some are not detected by IDE introspection (e.g. vertices).

**Describe the solution you'd like**
Add typing to `scipy/spatial/qhull.py` *-or-* add stubs if the `qhull.py` heavily depends on C code.

**Describe alternatives you've considered**
Add typing stubs outside of scipy - that would be time-consuming and hard to maintain if spatial (or qhull) change or add API.

**Additional context (e.g. screenshots)**
(none)

### PR DESCRIPTION
**Pull request by mckib2**
#### Reference issue

Closes https://github.com/scipy/scipy/issues/13299: static typing for qhull.pyx

#### What does this implement/fix?

- Add `qhull.pyi` stub file to declare types for classes/functions in `qhull.pyx`
- imports numpy.typing if possible, otherwise `numpy.typing.ArrayLike` falls back to `Any`

#### Additional information

- Unable to import `__future__ annotations` in Cython, so we need stub files
- I tried to type all functions that were possible, the only thing I think I missed was the `qhull_options` argument of the constructor for `_QhullUser`s because it has dynamic type during runtime (str and bytes)
- `mypy` appears to parse the file:
```python
import numpy as np
from scipy.spatial import ConvexHull
ConvexHull([]).add_points(np.random.rand(30, 2), 'not a bool')

# mypy output: test.py:3: error: Argument 2 to "add_points" of "ConvexHull" has incompatible type "str"; expected "bool"  [arg-type]
```

### CODE REVIEW

**Comment by rgommers** [comment]
Interesting, thanks @mckib2 

> I'm not sure if these annotations make it to IDEs or not

Trying with mypy on some example code would be the easiest - if mypy sees the annotations, then IDEs should work too. Anyway, mypy is the most important consumer, if it doesn't see these annotations inside `.pyx` files, then they won't be very useful and we need stub files instead.

**Comment by mckib2** [comment]
Looking at this a little bit more, it looks like mypy has no way of getting any info from Cython sources, so stubs it is.  Even though it looked like Cython was using some of the extra typing information, I'll go ahead and leave `qhull.pyx` unchanged.

My remaining concern is about how to keep the stubs and sources synced.  Is there a way we could test to make sure that the stubs were current?  If this isn't easy, it doesn't seem like that's in the scope of this PR, just a thought

**Comment by rgommers** [comment]
> My remaining concern is about how to keep the stubs and sources synced. Is there a way we could test to make sure that the stubs were current? If this isn't easy, it doesn't seem like that's in the scope of this PR, just a thought

Should be relatively straightforward, I hope. First step is to remove 
```
[mypy-scipy.spatial.qhull]
ignore_missing_imports = True
```
from `mypy.ini`, then run the tests with `python runtests.py --mypy` (see gh-11706).

You may see other failures though, because I think we haven't added this in CI - if so and it's more than a few, let's open a separate issue for that and fix it up. Now that numpy 1.20 has decent typing, we can get back to adding annotations in scipy much more easily.

**Comment by mckib2** [comment]
Thanks for pointing out mypy tests!

There are more than a few errors when I run (against `numpy==1.20.0rc2`):
```
Found 97 errors in 22 files (checked 647 source files)
```

I'll open another PR to see what we can do about those error.

File: scipy/spatial/qhull.pyi [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+if TYPE_CHECKING:
+    from numpy.typing import ArrayLike
+else:
+    ArrayLike = Any
+
+import numpy as np
+
+class _Qhull:
+    def __init__(self, points: ArrayLike, furthest_site: bool = ...,
+                 incremental: bool = ..., qhull_options=...): ...
```
Comment (person142) [inline scipy/spatial/qhull.pyi:16]: `qhull_options` should be `str` I think.

File: scipy/spatial/qhull.pyi [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+if TYPE_CHECKING:
+    from numpy.typing import ArrayLike
+else:
+    ArrayLike = Any
+
+import numpy as np
+
+class _Qhull:
+    def __init__(self, points: ArrayLike, furthest_site: bool = ...,
```
Comment (person142) [inline scipy/spatial/qhull.pyi:15]: Recommend formatting in the way we discussed in https://github.com/scipy/scipy/pull/13331#pullrequestreview-560737179.

**Review by person142 — APPROVED** [review]

File: scipy/spatial/qhull.pyi (hunk 1 above)
Comment (mckib2) [reply scipy/spatial/qhull.pyi:16]: I did try `str`, but it's converted later to `bytes` which mypy will complain about and I'm not mypy-savy enough to figure out how to get it working

File: scipy/spatial/qhull.pyi (hunk 2 above)
Comment (mckib2) [reply scipy/spatial/qhull.pyi]: Done

**Pull request merged**